### ISSUE TITLE
🐛 Source Stripe: New `events` based incremental sync doesn't capture updates

### ISSUE THREAD

**Issue by guillaume-chech** [airbytehq/airbyte#31917]
### Connector Name

stripe

### Connector Version

4.5.0

### What step the error happened?

During the sync

### Relevant information

I was thrilled to use the new incremental sync for Stripe connector to avoid having to use loopback mechanism to capture updates on objects given that Stripe API is reallllly slow. 

So I upgraded from 3.17.4 to 4.5.0, btw the migration guide is not really helpful and I had to : 
- Delete previous source and connection
- Create new source 
- Create a new connection

Anyway, with that done I made some test using the `_airbyte_raw_` tables to check incremental updates in the hope that for a given `payment_intents` i will get many rows, at least one per `status` update at least. 
However, I don't seem to get any update, while I do get multiple `events` for a payment intent in the `events` stream. 
This is the same for all objects with a status that support event based incremental update mechanism.

I can't give you our company data for reference but the steps to reproduce could something like this in a dev env :
- Create a payment intent
-> Run a sync and obeserve that 1 events and 1 payment_intent is replicated
- Capture the Charge for that payment intent
-> Run a sync and observe multiples `events` being replicated in the `events` stream but only one row in the `payment_intents` and `charges` stream remain in target tables with the original status.

This a pretty pretty big problem as now the streams that support  incremental updates do not apply loopback window, in the end we had to rollback to 3.17.4 in order to get up to date data. 

### Contribute

- [ ] Yes, I want to contribute

**Comment by michelgalle** [comment]
this might have to do with a problem that I am facing as well. with my troubleshooting I found that rows are ingested out of order into raw tables so when dedup is applied the last row used to generate the final table is not actually the most recent row for a given object. To check whether it is the same problem, query raw events table, filter by the object you want and check whether the order of events is the same when ordering by event_created_at and when ordering by airbyte_emitted_at column. in my case when I order by emitted_at column, the most recent row is not the last event for a given object.

**Comment by davydov-d** [comment]
> this might have to do with a problem that I am facing as well. with my troubleshooting I found that rows are ingested out of order into raw tables so when dedup is applied the last row used to generate the final table is not actually the most recent row for a given object. To check whether it is the same problem, query raw events table, filter by the object you want and check whether the order of events is the same when ordering by event_created_at and when ordering by airbyte_emitted_at column. in my case when I order by emitted_at column, the most recent row is not the last event for a given object.

:wave: hi @michelgalle thanks for your feedback.
The sorting order of the `emitted_at` column is not obliged to be the same as the one of `created_at` or `updated_at` fields. The Stripe API returns records in the descending order, that's why you can witness this kind of inconsistency. However, to make the connection work properly, a record should have a correct cursor value, which is, in most cases `updated_at` or `created_at`. Only in case two records have the same cursor value, the `emitted_at` field is considered. Please let me know which streams are you using? And what is the exact sync mode?

**Comment by michelgalle** [comment]
Hi @davydov-d,
Thanks for the update, I saw the issue in `subscriptions` and `invoices`, I am actually normalizing data from raw tables directly, airbyte normalization generates extra tables that uses space and the normalization step is not that performant (postgres). upon checking the data, the problem is that once invoice is finalized, it has around 4 events per invoice at the same timestamp of invoice finalized so they all have the same updated_at timestamp. and the last one of those 4, if looking at emitted_at is not the most up to date event.

**Comment by davydov-d** [comment]
> Hi @davydov-d, Thanks for the update, I saw the issue in `subscriptions` and `invoices`, I am actually normalizing data from raw tables directly, airbyte normalization generates extra tables that uses space and the normalization step is not that performant (postgres). upon checking the data, the problem is that once invoice is finalized, it has around 4 events per invoice at the same timestamp of invoice finalized so they all have the same updated_at timestamp. and the last one of those 4, if looking at emitted_at is not the most up to date event.

Thanks, now I understand the problem. We'll try to figure something out, will keep you updated

**Comment by davydov-d** [comment]
Hi @guillaume-chech and thanks for your feedback. 
I have just tried to reproduce the issue you are facing but I did not succeed. Perhaps I am missing something. In order to reproduce it step by step, I'd like you to provide me with more details please:
 - What is the delay between capturing the charge and running an incremental sync?
 - Are you capturing it via API or UI?
 - What are the exact types of events you're getting in a subsequent sync in the Events stream and not getting in the Payment Intents stream?
 - Does the value of the `updated_at` field of the payment intent change after the subsequent sync? 
 - Can you please share the log of the sync that is not producing expected records?
 
 Thanks in advance!

**Comment by guillaume-chech** [comment]
Hi @davydov-d thanks a lot for looking into it ! 

Actually the issue seems to have resolved since then. I don't exactly get why but I expect the issue to be the one that @michelgalle reported since I'm also using raw data and custom normalization.

However I do have some issue with the `refunds` stream. 
On my Connection I have a state : 
```
  {
    "streamDescriptor": {
      "name": "refunds"
    },
    "streamState": {
      "updated": 1698332986
    }
  }
```
Which corresponds to 2023-10-26 roughly. 
The connection is set to run using incremental-append with key id and cursor updated_at.
On subsequent runs, I always get 0 records returned , yet I know for sure that refunds were created. One of the reason being i do get events of type `charge.refunded` returned by the same connection for instanct.

I checked on my stripe account and I don't have any event of type `refund.created` or `refund.updated` , which is the events used to get updates but they are not used on our instance, since i expect they corresponds to refunds being created via CLI on Stripe, most refunds being created as part of other ressources lifecycle.

However, we should get new records by calling the refunds endpoint using created param set to the state updated value. 

So in order to investigate further I ran the latest version of the stripe source locally with --debug, set the state to the one in production and checked what happened, especially API calls . 

I would have expected the following :
- 1 call for refunds endpoint with created param set to updated value of the state file
- 1 call for events endpoint with created param set to updated value of the state file and type param set to refund.created or refund.updated

Instead I got : 
- 1 call for refunds endpoint that did retrieve records
- No call to events endpoint

However, somehow records retrieved from the refunds endpoint were not taken into account for some reason. 

You'll find attached the logs of the local run for reference and the log of the production run.
[production_logs.txt](https://github.com/airbytehq/airbyte/files/13243267/production_logs.txt)
[local_logs.json](https://github.com/airbytehq/airbyte/files/13243276/local_logs.json)

So my questions boil down to : 

1. Why records from `refunds` endpoints are ignored ?
2. (Optional) Why don't we call the `events` endpoint for type `refund.created` or `refund.updated` as the code would imply ?

NB : I'm running 4.5.1 on both prod and local.

**Comment by davydov-d** [comment]
@guillaume-chech the Refunds stream issue is known and will be fixed as part of [#3428](https://github.com/airbytehq/oncall/issues/3428#issuecomment-1789175659) in the nearest future. As of now our incremental streams do not support regular endpoints + event-based approach. We use either of them depending on the selected sync mode and state value. Therefore we will use regular endpoint to sync refunds and filter them by the `created` field. To have your refunds always up-to-date you'll be able to make use of the lookback window.

Regarding the problem @michelgalle reported, unfortunately I can not take the responsibility to decide on which of the records with the same ID and `updated` value is the up-to-date based on just the event type. The root cause of this problem is the Stripe API generating multiple events with the same timestamp which results in multiple records with different state that got deduplicated based on the time they were emitted.

In order to work this around I can recommend you use the Incremental + Append mode and make custom deduplication.

I'm closing this issue now, please feel free to contact me in case you have any other questions
